Bench FjKmeans fails with java.lang.NullPointerException

repro command : java.exe -Xms4G -Xmx4G -debug -jar renaissance-jmh-0.14.1-10.jar -bm avgt -prof stack  -f 1 -t 20 -i 20 -v EXTRA FjKmeans

JDK versions: Microsoft open JDK build jdk-17.0.6+10

Exception:
java.lang.NullPointerException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:564)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:591)
	at java.base/java.util.concurrent.ForkJoinTask.joinForPoolInvoke(ForkJoinTask.java:1042)
	at java.base/java.util.concurrent.ForkJoinPool.invoke(ForkJoinPool.java:2639)
	at org.renaissance.jdk.concurrent.JavaKMeans.run(JavaKMeans.java:54)
	at org.renaissance.jdk.concurrent.FjKmeans.$anonfun$run$1(FjKmeans.scala:51)
	at org.renaissance.jdk.concurrent.FjKmeans.$anonfun$run$1$adapted(FjKmeans.scala:50)
	at scala.collection.immutable.Range.map(Range.scala:59)
	at org.renaissance.jdk.concurrent.FjKmeans.run(FjKmeans.scala:50)
	at org.renaissance.jmh.JmhRenaissanceBenchmark.runOperation(JmhRenaissanceBenchmark.java:128)
	at org.renaissance.jdk.concurrent.jmh_generated.JmhFjKmeans_runOperation_jmhTest.runOperation_AverageTime(JmhFjKmeans_runOperation_jmhTest.java:226)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:475)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:458)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.NullPointerException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:564)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:591)
	at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:672)
	at org.renaissance.jdk.concurrent.JavaKMeans$RangedTask.compute(JavaKMeans.java:149)
	at java.base/java.util.concurrent.RecursiveTask.exec(RecursiveTask.java:100)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.concurrent.ForkJoinPool.invoke(java.util.concurrent.ForkJoinTask)" because the return value of "org.renaissance.jdk.concurrent.JavaKMeans$UpdateTask.getPool()" is null
	at org.renaissance.jdk.concurrent.JavaKMeans$UpdateTask.average(JavaKMeans.java:330)
	at org.renaissance.jdk.concurrent.JavaKMeans$UpdateTask.computeClusterAverages(JavaKMeans.java:315)
	at org.renaissance.jdk.concurrent.JavaKMeans$UpdateTask.computeDirectly(JavaKMeans.java:306)
	at org.renaissance.jdk.concurrent.JavaKMeans$UpdateTask.computeDirectly(JavaKMeans.java:273)
	at org.renaissance.jdk.concurrent.JavaKMeans$RangedTask.compute(JavaKMeans.java:143)
	at java.base/java.util.concurrent.RecursiveTask.exec(RecursiveTask.java:100)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinTask.awaitDone(ForkJoinTask.java:436)
	at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:670)
	at org.renaissance.jdk.concurrent.JavaKMeans$RangedTask.compute(JavaKMeans.java:149)
	at java.base/java.util.concurrent.RecursiveTask.exec(RecursiveTask.java:100)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool.externalHelpQuiescePool(ForkJoinPool.java:2104)
	at java.base/java.util.concurrent.ForkJoinPool.awaitQuiescence(ForkJoinPool.java:3321)
	at org.renaissance.jdk.concurrent.JavaKMeans.run(JavaKMeans.java:59)
	at org.renaissance.jdk.concurrent.FjKmeans.$anonfun$run$1(FjKmeans.scala:51)
	at org.renaissance.jdk.concurrent.FjKmeans.$anonfun$run$1$adapted(FjKmeans.scala:50)
	at scala.collection.immutable.Range.map(Range.scala:59)
	at org.renaissance.jdk.concurrent.FjKmeans.run(FjKmeans.scala:50)
	at org.renaissance.jmh.JmhRenaissanceBenchmark.runOperation(JmhRenaissanceBenchmark.java:128)
	at org.renaissance.jdk.concurrent.jmh_generated.JmhFjKmeans_runOperation_jmhTest.runOperation_AverageTime(JmhFjKmeans_runOperation_jmhTest.java:226)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:475)
	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:458)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

--- a/benchmarks/jdk-concurrent/src/main/java/org/renaissance/jdk/concurrent/JavaKMeans.java
+++ b/benchmarks/jdk-concurrent/src/main/java/org/renaissance/jdk/concurrent/JavaKMeans.java
@@ -47,17 +47,20 @@ public final class JavaKMeans {
   public List<Double[]> run(
     final int clusterCount, final List <Double[]> data, final int iterationCount
   ) throws InterruptedException, ExecutionException {
-    List<Double[]> centroids = randomSample(clusterCount, data, new Random(100));
-    for (int iteration = 0; iteration < iterationCount; iteration++) {
-      final AssignmentTask assignmentTask = new AssignmentTask(data, centroids);
-      final UpdateTask updateTask = new UpdateTask(forkJoin.invoke(assignmentTask));
-      final Map<Double[], List<Double[]>> clusters = forkJoin.invoke(updateTask);
+    // Submit the k-means computation as a task to the fork-join pool to
+    // ensure that forked tasks are executed by worker threads from the pool.
+    return forkJoin.submit(() -> {
+      List<Double[]> centroids = randomSample(clusterCount, data, new Random(100));
+      for (int iteration = 0; iteration < iterationCount; iteration++) {
+        final AssignmentTask assignmentTask = new AssignmentTask(data, centroids);
+        final UpdateTask updateTask = new UpdateTask(assignmentTask.invoke());
+        final Map<Double[], List<Double[]>> clusters = updateTask.invoke();
 
-      centroids = new ArrayList<>(clusters.keySet());
-    }
+        centroids = new ArrayList<>(clusters.keySet());
+      }
 
-    forkJoin.awaitQuiescence(1, TimeUnit.SECONDS);
-    return centroids;
+      return centroids;
+    }).get();
   }
 
 

--- a/benchmarks/jdk-concurrent/src/main/java/org/renaissance/jdk/concurrent/JavaKMeans.java
+++ b/benchmarks/jdk-concurrent/src/main/java/org/renaissance/jdk/concurrent/JavaKMeans.java
@@ -327,7 +327,7 @@ public final class JavaKMeans {
 
     private double[] average(final List<Double[]> elements) {
       final VectorSumTask sumTask = new VectorSumTask(elements);
-      final double[] vectorSums = getPool().invoke(sumTask);
+      final double[] vectorSums = sumTask.invoke();
       return div(vectorSums, elements.size());
     }
 

